src/guiscalingfilter.cpp

```diff
@@ -0,0 +1,154 @@
+#include "guiscalingfilter.h"
+#include "imagefilters.h"
+#include "settings.h"
+#include "main.h"		// for g_settings
+#include <stdio.h>
+
+/* Maintain a static cache to store the images that correspond to textures
+ * in a format that's manipulable by code.  Some platforms exhibit issues
+ * converting textures back into images repeatedly, and some don't even
+ * allow it at all.
+ */
+std::map<io::path, video::IImage *> imgCache;
+
+/* Maintain a static cache of all pre-scaled textures.  These need to be
+ * cleared as well when the cached images.
+ */
+std::map<io::path, video::ITexture *> txrCache;
+
+/* Manually insert an image into the cache, useful to avoid texture-to-image
+ * conversion whenever we can intercept it.
+ */
+void guiScalingCache(io::path key, video::IVideoDriver *driver, video::IImage *value) {
+	if (!g_settings->getBool("gui_scaling_filter"))
+		return;
+	video::IImage *copied = driver->createImage(value->getColorFormat(),
+			value->getDimension());
+	value->copyTo(copied);
+	imgCache[key] = copied;
+}
+
+// Manually clear the cache, e.g. when switching to different worlds.
+void guiScalingCacheClear(video::IVideoDriver *driver) {
+	for (std::map<io::path, video::IImage *>::iterator it = imgCache.begin();
+			it != imgCache.end(); it++) {
+		if (it->second != NULL)
+			it->second->drop();
+	}
+	imgCache.clear();
+	for (std::map<io::path, video::ITexture *>::iterator it = txrCache.begin();
+			it != txrCache.end(); it++) {
+		if (it->second != NULL)
+			driver->removeTexture(it->second);
+	}
+	txrCache.clear();
+}
+
+#ifdef __ANDROID__
+// Compute next-higher power of 2 efficiently, for platforms that
+// require power-of-2 textures.
+u32 nextpower2(u32 orig) {
```

Comment:
This would be better in utils/numeric (probably as an inline function).  The function is generic, doesn't just work on Android and maybe someone else will use it one day.  They won't use it again one day if it's hidden away in this file :)